### PR TITLE
Initial implementation of virtfs-ext3

### PR DESCRIPTION
**Pull request by movitto**
Includes sequential modifications to existing codebase to:

- bring it up to standards
- implement the virtfs interface
- add specs

### CODE REVIEW

**Comment by movitto** [comment]
@roliveri this should be ready for review

**Comment by roliveri** [comment]
@hsong-rh @jerryk55 Please join in on the review.

File: lib/virtfs/ext3/directory.rb [hunk 1]
```diff
@@ -1,51 +1,71 @@
...
-      @sb = sb; @inodeNum = inodeNum
-      @inodeObj = sb.getInode(inodeNum)
-      @data = FileData.new(@inodeObj, @sb).read
+    def initialize(sb, inode_num = ROOT_DIRECTORY)
+      raise "nil superblock"   if sb.nil?
+      raise "nil inode number" if inode_num.nil?
+      @sb        = sb
+      @inode_num = inode_num
+      @inode_obj = sb.get_inode(inode_num)
+      @data      = sb.ext4? ? @inode_obj.read : FileData.new(@inode_obj, @sb).read
```
Comment (jerryk55) [inline lib/virtfs/ext3/directory.rb]: I'm curious as to why we would continue handling this is an Ext3 filesystem if it is actually an Ext4 filesystem.  Shouldn't this be an error and/or an exception?

File: lib/virtfs/ext3/directory.rb [hunk 2]
```diff
@@ -1,51 +1,71 @@
...
 
-    def initialize(sb, inodeNum = ROOT_DIRECTORY)
-      raise "Ext3::Directory.initialize: Nil superblock"   if sb.nil?
-      raise "Ext3::Directory.initialize: Nil inode number" if inodeNum.nil?
-      @sb = sb; @inodeNum = inodeNum
-      @inodeObj = sb.getInode(inodeNum)
-      @data = FileData.new(@inodeObj, @sb).read
+    def initialize(sb, inode_num = ROOT_DIRECTORY)
+      raise "nil superblock"   if sb.nil?
+      raise "nil inode number" if inode_num.nil?
```
Comment (jerryk55) [inline lib/virtfs/ext3/directory.rb]: Is the assumption here that the raised exception will display the Module::Class.method info?  Why else would you remove this helpful info?

File: lib/virtfs/ext3/alloc_bitmap.rb [hunk 3]
```diff
@@ -1,16 +1,16 @@
-module Ext3
+module VirtFS::Ext3
   class AllocBitmap
     def initialize(data)
-      raise "Ext3::AllocBitmap.initialize: Nil data" if data.nil?
+      raise "Ext::AllocBitmap.initialize: Nil data" if data.nil?
```
Comment (jerryk55) [inline lib/virtfs/ext3/alloc_bitmap.rb]: Why did you remove the "3"?

File: lib/virtfs/ext3/directory.rb [hunk 4]
```diff
@@ -1,51 +1,85 @@
...
+
+    def initialize(fs, sb, inode_num = ROOT_DIRECTORY)
+      raise "nil superblock"   if sb.nil?
+      raise "nil inode number" if inode_num.nil?
+      @fs        = fs
+      @sb        = sb
+      @inode_num = inode_num
+      @inode_obj = sb.get_inode(inode_num)
+      @data      = sb.ext4? ? @inode_obj.read : FileData.new(@inode_obj, @sb).read
+    end
```
Comment (jerryk55) [inline lib/virtfs/ext3/directory.rb:18]: If the superblock is Ext4, isn't that an error or an exception?  I wouldn't think we would just continue to run the initialization as if nothing happened.

File: lib/virtfs/ext3/directory.rb [hunk 5]
```diff
@@ -1,51 +1,85 @@
...
       @ents_by_name = {}; p = 0
       return @ents_by_name if @data.nil?
-      newEnt = @sb.isNewDirEnt?
+      new_entry = @sb.new_dir_entry?
       loop do
         break if p > @data.length - 4
         break if @data[p, 4].nil?
-        de = DirectoryEntry.new(@data[p..-1], newEnt)
-        raise "Ext3::Directory.globEntries: DirectoryEntry length cannot be 0" if de.len == 0
+        de = DirectoryEntry.new(fs, @data[p..-1], new_entry)
```
Comment (jerryk55) [inline lib/virtfs/ext3/directory.rb]: Why did you add a new "fs" argument?  It is not used in the DirectoryEntry class except to save it, and I can find no other code that makes use of the accessor method.

File: lib/virtfs/ext3/directory_entry.rb [hunk 6]
```diff
@@ -31,37 +31,65 @@ class DirectoryEntry
...
+      Time.now
     end
 
-    def dump
-      out = "\#<#{self.class}:0x#{'%08x' % object_id}>\n"
-      out += "Inode   : #{inode}\n"
-      out += "Len     : #{len}\n"
-      out += "Name len: 0x#{'%04x' % @de['name_len']}\n"
-      out += "Type    : #{fileType}\n" if @isNew
-      out += "Name    : #{name}\n"
```
Comment (jerryk55) [inline lib/virtfs/ext3/directory_entry.rb:64]: The dump methods are very helpful when attempting to debug SSA issues. I would lobby to put them back (I haven't noticed if you removed any other ones).

File: lib/virtfs/ext3/file_data.rb [hunk 7]
```diff
@@ -68,17 +68,17 @@ def write(buf, _len = buf.length)
     private
 
     def getStartBlock(pos, len)
-      startBlock, startByte = pos.divmod(@blockSize)
-      endBlock, endByte = (pos + len - 1).divmod(@blockSize)
+      startBlock, startByte = pos.divmod(@block_size)
+      endBlock, endByte = (pos + len - 1).divmod(@block_size)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_data.rb]: If you're going through the incredibly tedious chore of aligning things and getting rid of camel-case when possible maybe the "=" here should line up with the previous line?

File: lib/virtfs/ext3/group_descriptor_entry.rb [hunk 8]
```diff
@@ -31,35 +31,20 @@ def initialize(buf)
...
-    def inodeTable
+    def inode_table
       @gde['inode_table']
     end
 
-    def numDirs
+    def num_dirs
       @gde['num_dirs']
     end
-
```
Comment (jerryk55) [inline lib/virtfs/ext3/group_descriptor_entry.rb]: See previous comment about dump methods.

File: lib/virtfs/ext3/group_descriptor_table.rb [hunk 9]
```diff
@@ -32,23 +32,11 @@ def [](group)
       @gdt[group]
     end
 
-    def dump(dump_bitmaps = false)
```
Comment (jerryk55) [inline lib/virtfs/ext3/group_descriptor_table.rb]: Ditto.

**Review by jerryk55 — COMMENTED** [review]
@movitto a lot to go through.  I"m going to add some comments now but continue with more tomorrow...

**Comment by jerryk55** [comment]
@movitto it looks like the tests are failing because something can't find virtfs.

File: lib/virtfs/ext3/superblock.rb [hunk 10]
```diff
@@ -208,199 +208,111 @@ def gdt
...
-
     # ////////////////////////////////////////////////////////////////////////////
     # // Utility functions.
 
-    def gotBit?(field, bit)
+    def bit?(field, bit)
       field & bit == bit
     end
-
-    # Dump object.
```
Comment (jerryk55) [inline lib/virtfs/ext3/superblock.rb]: Yet another ditto to the previous comments on dump.  You get the picture.  :-)

File: lib/virtfs/ext3/file_class.rb [hunk 11]
```diff
@@ -0,0 +1,188 @@
+module VirtFS::Ext3
+  class FS
+    def file_atime(p)
+      f = get_file(p)
+      raise Errno::ENOENT, "No such file or directory" if f.nil?
+      f.atime
+    end
+
+    def file_blockdev?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+      VirtFS::Stat.new(VirtFS::Ext3::File.new(file, superblock).to_h)
+    end
+
+    def file_mtime(p)
+      f = get_file(p)
+      raise Errno::ENOENT, "No such file or directory" if f.nil?
+      f.mtime
+    end
+
+    def file_owned?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Several more unimplemented methods here?

File: lib/virtfs/ext3/file_class.rb [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+      !f.nil? && f.file?
+    end
+
+    def file_ftype(p)
+    end
+
+    def file_grpowned?(p)
+    end
+
+    def file_identical?(p1, p2)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+    def file_file?(p)
+      f = get_file(p)
+      !f.nil? && f.file?
+    end
+
+    def file_ftype(p)
+    end
+
+    def file_grpowned?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    def file_exist?(p)
+      !get_file(p).nil?
+    end
+
+    def file_file?(p)
+      f = get_file(p)
+      !f.nil? && f.file?
+    end
+
+    def file_ftype(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+    def file_directory?(p)
+      f = get_file(p)
+      !f.nil? && f.dir?
+    end
+
+    def file_executable?(p)
+    end
+
+    def file_executable_real?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    def file_delete(p)
+      raise "writes not supported"
+    end
+
+    def file_directory?(p)
+      f = get_file(p)
+      !f.nil? && f.dir?
+    end
+
+    def file_executable?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/file_class.rb [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    def file_atime(p)
+      f = get_file(p)
+      raise Errno::ENOENT, "No such file or directory" if f.nil?
+      f.atime
+    end
+
+    def file_blockdev?(p)
+    end
+
+    def file_chardev?(p)
```
Comment (jerryk55) [inline lib/virtfs/ext3/file_class.rb]: Not yet implemented?

File: lib/virtfs/ext3/directory.rb (hunk 5 above)
Comment (movitto) [reply lib/virtfs/ext3/directory.rb]: The 'fs' argument was added to comply with the 'virtfs' interface which is expecting it in that accessor on Directory instances. Without it, virtfs will throw an error when it tries to call the 'fs' method

File: lib/virtfs/ext3/alloc_bitmap.rb (hunk 3 above)
Comment (movitto) [reply lib/virtfs/ext3/alloc_bitmap.rb]: I was thinking so that there errors would be more generic / applicable to both the Ext3 and Ext4 modules. Not a big deal, can change it back if that is preferred

File: lib/virtfs/ext3/directory_entry.rb (hunk 6 above)
Comment (movitto) [reply lib/virtfs/ext3/directory_entry.rb:64]: As mentioned earlier, the dump methods were removed at the same time as the references to $log. I'll readd the dump methods with the next PR revision, though we need add a central logging mechanism to the virtfs module

File: lib/virtfs/ext3/file_class.rb (hunk 11 above)
Comment (movitto) [reply lib/virtfs/ext3/file_class.rb]: Yes, a few features in the new virtfs interface were never implemented in MiQFS, since these pull requests are for extraction / refactoring / testing, I did not make any functional changes / add new features

**Comment by movitto** [comment]
@jerryk55 see comments above (updated PR will be sent tomorrow)

File: lib/virtfs/ext3/directory.rb (hunk 4 above)
Comment (movitto) [reply lib/virtfs/ext3/directory.rb:18]: @jerryk55 the Ext4::Inode class provides a 'read' method which we can use to access the inode data directly, but not such method exists in the ext3 module. This was added so that both modules can leverage the common Directory code.

At some point the "read" method should be added to the Ext3::inode class for consistency.

**Comment by movitto** [comment]
@jerryk55 @roliveri pushed the updated PR, a few changes

- the patchset has been simplied to 1. refactoring 2. virtfs interface 3. specs

- the 'dump' methods were readded and other outstanding comments addressed

- more style changes were added for consistency

**Comment by movitto** [comment]
As a side note, we discovered a potential bug w/ the current ext3 module today that needs further exploration (and potential fixing both here and in the main miq codebase):

At line 93 of the current "file_data.rb" module, we seemingly retrieve the triple indirect block using the double indirect pointer. This needs investigating & a possible fix.

https://github.com/ManageIQ/virtfs-ext3/blob/master/lib/virtfs/ext3/file_data.rb#L99

File: spec/factories/ext.rb [hunk 19]
```diff
@@ -0,0 +1,13 @@
+require 'ostruct'
+require 'virtfs/block_io'
+require 'virt_disk/block_file'
+
+FactoryGirl.define do
+  factory :ext, class: OpenStruct do
+    path '/home/mmorsi/workspace/cfme/virtfs-ext3/ext3.fs'
```
Comment (jerryk55) [inline spec/factories/ext.rb]: Just noticed this path.  Obviously you want something else here...

File: lib/virtfs/ext3/base.rb [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        VirtFS::Ext3::Superblock.new(blk_device)
+        return true
+      rescue => err
+        return false
+      end
+    end
+
+    def initialize(blk_device)
+      blk_device.seek(0, IO::SEEK_SET)
+      @superblock  = Superblock.new(blk_device)
```
Comment (jerryk55) [inline lib/virtfs/ext3/base.rb]: Just noticed that you invoke "Superblock.new" here but on line 12 above the call is "VirtFS::Ext3::Superblock.new".
If this line works then the more explicit line is probably not necessary, right?

File: lib/virtfs/ext3.rb [hunk 21]
```diff
@@ -1 +1,16 @@
 require "virtfs/ext3/version"
+require_relative 'ext3/alloc_bitmap'
```
Comment (jerryk55) [inline lib/virtfs/ext3.rb:2]: See @Fryguy comment on my PR https://github.com/ManageIQ/virtfs-xfs/pull/1 - "Minor, but it's common in gems to just refer to file assuming lib is on the path. That is, instead of require_relative 'xfs/whatever', prefer require 'virtfs/xfs/whatever'".  If this is desired we should be consistent throughout all the VirtFS and VirtFS-fstype gems.

File: lib/virtfs/ext3/directory_entry.rb (hunk 6 above)
Comment (movitto) [reply lib/virtfs/ext3/directory_entry.rb:64]: @jerryk55 dump methods readded & standardized

File: lib/virtfs/ext3/base.rb (hunk 20 above)
Comment (movitto) [reply lib/virtfs/ext3/base.rb]: @jerryk55 simplied w/ the new update

File: lib/virtfs/ext3.rb (hunk 21 above)
Comment (movitto) [reply lib/virtfs/ext3.rb:2]: @jerryk55 @Fryguy reading through the xfs PR, it seems that we are going w/ 'require_relative' correct? And thus no additional work is needed for this

**Comment by movitto** [comment]
@roliveri @jerryk55 this should be pretty close to go. That being said the camcorderfs integration still isn't working 100% as intended. The recorder is wired up and the cassettes written but I don't believe the right transactions are being recorded & playback is failing ('disk not found' once it is moved).

Does anything look off w/ the initialization code? (in spec_helper)

File: lib/virtfs/ext3/superblock.rb [hunk 22]
```diff
@@ -137,270 +139,309 @@ class Superblock
...
       @sb['last_mnt_path'].delete!("\000")
-      @numGroups, @lastGroupBlocks = @sb['num_blocks'].divmod(@sb['blocks_in_group'])
-      @numGroups += 1 if @lastGroupBlocks > 0
-      @fsId = MiqUUID.parse_raw(@sb['fs_id'])
-      @volName = @sb['vol_name']
-      @jrnlId = MiqUUID.parse_raw(@sb['jrnl_id'])
+      @num_groups, @last_group_blocks = @sb['num_blocks'].divmod(@sb['blocks_in_group'])
+
+      @num_groups += 1 if @last_group_blocks > 0
+      @fsId        = DiskUUID.parse_raw(@sb['fs_id'])
```
Comment (jerryk55) [inline lib/virtfs/ext3/superblock.rb:203]: I assume that DiskUUID isn't really defined anywhere?

File: lib/virtfs/ext3/superblock.rb [hunk 23]
```diff
@@ -1,19 +1,16 @@
 # encoding: US-ASCII
 
-require 'fs/ext3/group_descriptor_table'
-require 'fs/ext3/inode'
+require_relative 'group_descriptor_table'
+require_relative 'inode'
 
 require 'binary_struct'
-require 'util/miq-uuid'
+require 'virt_disk/disk_uuid'
```
Comment (jerryk55) [inline lib/virtfs/ext3/superblock.rb:7]: This file doesn't exist, right?  I noticed that you added a dependency to the UUIDTools gem in the Gemfile - is your intent to change this to use that other gem?

File: Gemfile [hunk 24]
```diff
@@ -1,5 +1,8 @@
 source 'https://rubygems.org'
 
+gem 'virtfs', "~> 0.0.1", :git => "https://github.com/ManageIQ/virtfs.git", :branch => "master"
+gem 'comcorderfs', "~> 0.1.0", :git => "https://github.com/ManageIQ/virtfs-camcorderfs.git", :branch => "master"
```
Comment (jerryk55) [inline Gemfile]: Typo in 'comcorderfs', right?

**Comment by movitto** [comment]
@jerryk55 ah yes, I had those components in my virt_disk checkout locally but never committed them. Updated virt_disk PR #1 to include them.

Also fixed the typo, good catch!

**Comment by roliveri** [comment]
@movitto I think I might know the cause of your test issues with CamcorderFS.
The tests don't activate VirtFS, explicitly referencing `VirtFS::VFile, etc`, but I'm not sure what `VirtDisk::BlockFile` is doing. If it's calling `File.open` instead of `VirtFS::Vfile.open` then the IO from the disk file won't be recorded.

If that's the case, there are 2 possible solutions:
1 - Activate `VirtFS` in the tests.
2 - Change the version of `BlockFile` that the tests use to use `VirtFS::VFile`.

Option #2 would probably be the easiest solution. I'd just create a test version of `BlockFile` in the spec directory and use that.

**Comment by movitto** [comment]
@roliveri I just updated the virt_disk PR to check if VirtFS is defined and if so use VirtFS::VFile, else use the existing File.open mechanism. Unfortunately while some progress was made (some tests broke due to the referencing of an incorrect mount point which I fixed), the cassettes still do not contain the relevant recordings.

Pushed the updated virtfs-ext3 and virt_disk pr in the hopes of getting more eyeballs on this to sort it out.

**Comment by movitto** [comment]
@roliveri also I checked and I believe the virtfs-ext3 module is compliant w/ how we are approaching the project layout for these components. If I'm wrong or missed anything just shout out and I can fix

**Comment by movitto** [comment]
@roliveri @jerryk55  success! The addition to virt_disk/block_file was needed as per your suggestion rich, but I had to do the opposite of your other suggestion and leave set the camcorder recorder root when mounting that device (perhaps this should be done internally in camcorderfs). 

https://github.com/ManageIQ/virtfs-ext3/pull/1/commits/7346f2c9fd3595138ed551508609560c30161dc1#diff-23dc0ec957f4bb1eba766196149f8004R19

I noticed that when it wasn't there more disk interactions were being recorded (as a result of the addition to virt_disk) but they were incorrect. After this change camcorder was able to pickup all the operations on the ext3 image being done through virtfs.

Small fix but took a while to discover. Good thing is though I verified that the specs fully run from the cassettes when the disk image is moved, proving the camcorderfs concept, and providing an example for further implementations.

Should be just about ready to merge!

**Comment by roliveri** [comment]
@movitto Ah, I forgot I suggested removing setting root - yeah, that was just an initial shot in the dark, and was wrong. Sorry my suggestion caused you to waste so much time.

**Comment by roliveri** [comment]
Now the question is, why can't the test find camcorders?

**Comment by movitto** [comment]
@roliveri @jerryk55 the camcorder based tests are now working and this PR should be ready to go. Besides various issues w/ deps I had to add a fixed path to the factory used in the specs so that travis could run against consistent targets. We may want to change this at some point, but other than that this is done

File: lib/virtfs/ext3.rb (hunk 21 above)
Comment (roliveri) [reply lib/virtfs/ext3.rb:2]: @movitto Correct.

File: lib/virtfs/ext3/base.rb [hunk 25]
```diff
@@ -0,0 +1,50 @@
+module VirtFS::Ext3
+  class FS
```
Comment (roliveri) [inline lib/virtfs/ext3/base.rb]: Following the file naming convention, this file should be: `virtues/ext3/fs.rb`.

File: lib/virtfs/ext3/dir_class.rb [hunk 26]
```diff
@@ -0,0 +1,74 @@
+require 'active_support/core_ext/object/try' # until we can use the safe nav operator
+
+module VirtFS::Ext3
+  class FS
```
Comment (roliveri) [inline lib/virtfs/ext3/dir_class.rb]: Because re-opening the class to extend it breaks the file naming conventions, we've updated code like this to implement the methods in a module, which can then be included in the class in question.

So here, you'd implement the `DirClassMethods` module under the `FS` class:
```ruby
module VirtFS::Ext3
  class FS
    module DirCalssMethods
    :
    end
  end
end
```

and the file would then be: `virtfs/ext3/fs/dir_class_methods.rb`.
The class would then be included in the FS class.

File: lib/virtfs/ext3/file_class.rb [hunk 27]
```diff
@@ -0,0 +1,188 @@
+module VirtFS::Ext3
+  class FS
```
Comment (roliveri) [inline lib/virtfs/ext3/file_class.rb]: Same comment here as for the Dir class methods:

Because re-opening the class to extend it breaks the file naming conventions, we've updated code like this to implement the methods in a module, which can then be included in the class in question.

So here, you'd implement the `FileClassMethods` module under the `FS` class:
```ruby
module VirtFS::Ext3
  class FS
    module FileCalssMethods
    :
    end
  end
end
```

and the file would then be: `virtfs/ext3/fs/file_class_methods.rb`.
The class would then be included in the FS class.

File: lib/virtfs/ext3/file_class.rb [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+      raise Errno::ENOENT, "No such file or directory" if file.nil?
+      file
+    end
+
+    private
+
+      def get_file(p)
+        p = unnormalize_path(p)
+
+        dir, fname = ::File.split(p)
```
Comment (roliveri) [inline lib/virtfs/ext3/file_class.rb]: Remember under VirtFS the ::File and ::Dir classes can be reassigned. If you need to ensure you're using the core class, you should use `VfsRealFile`.

File: lib/virtfs/ext3/file_class.rb [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+    def file_world_writable?(p)
+    end
+
+    def file_writable?(p)
+    end
+
+    def file_writable_real?(p)
+    end
+
+    def file_new(f, parsed_args, _open_path, _cwd)
```
Comment (roliveri) [inline lib/virtfs/ext3/file_class.rb]: The `file_new` method needs to return a `File` class instance that implements the required file instance methods for the thin interface (see `virtfs/nativefs/thin/file.rb`). This appears to return a directory entry object, that doesn't appear to implement the required methods - so I'm not sure how this can work.

File: lib/virtfs/ext3/superblock.rb [hunk 30]
```diff
@@ -1,19 +1,16 @@
...
 require 'stringio'
 require 'memory_buffer'
 
 require 'rufus/lru'
 
-module Ext3
-  # ////////////////////////////////////////////////////////////////////////////
-  # // Data definitions. Linux 2.6.2 from Fedora Core 6.
-
+module VirtFS::Ext3
```
Comment (roliveri) [inline lib/virtfs/ext3/superblock.rb]: Should the following constants be defined under the `SuperBlock` class instead of dumping them in the `Ext3` class?

File: spec/dir_class_spec.rb [hunk 31]
```diff
@@ -0,0 +1,188 @@
...
+    "spec/cassettes/dir_class.yml"
+  end
+
+  describe ".[]" do
+    it "should return empty array when in a nonexistent directory" do
+      VirtFS.cwd = "/not_a_dir" # bypass existence checks.
+      expect(VirtFS::VDir["*"]).to match_array([])
+    end
+
+    it "should enumerate the same file names as the standard Dir.glob - simple glob" do
```
Comment (roliveri) [inline spec/dir_class_spec.rb:14]: I can see how tests like this work for the NativeFS modules, because the same files can be accessed through the standard interface as well as VirtFS, but how does this work for the Ext3 module? The files on the ext filesystem aren't accessible through the standard interface.

**Review by roliveri — CHANGES_REQUESTED** [review]

**Comment by movitto** [comment]
@roliveri  updated, see feedback above

File: lib/virtfs/ext3/dir_class.rb (hunk 26 above)
Comment (movitto) [reply lib/virtfs/ext3/dir_class.rb]: @roliveri updated

File: lib/virtfs/ext3/file_class.rb (hunk 27 above)
Comment (movitto) [reply lib/virtfs/ext3/file_class.rb]: @roliveri updated

File: lib/virtfs/ext3/file_class.rb (hunk 29 above)
Comment (movitto) [reply lib/virtfs/ext3/file_class.rb]: @roliveri ah this is because DirectoryEntry implements all the required methods, in the existing ext3 code which this was based off of, File just dispatches to that

https://github.com/ManageIQ/manageiq-gems-pending/blob/master/lib/gems/pending/fs/MiqFS/modules/Ext3.rb#L32

Regardless, I updated the patch to wrap the dir entry in a File instance and added methods to the File class delegating calls to the encapsulated DE

File: lib/virtfs/ext3/file_class.rb (hunk 28 above)
Comment (movitto) [reply lib/virtfs/ext3/file_class.rb]: @roliveri updated

File: lib/virtfs/ext3/superblock.rb (hunk 30 above)
Comment (movitto) [reply lib/virtfs/ext3/superblock.rb]: @roliveri  moved these under superblock

File: spec/dir_class_spec.rb (hunk 31 above)
Comment (movitto) [reply spec/dir_class_spec.rb:14]: @roliveri not fully following, these tests are testing access to the mounted filesystem (the actual mounting occurs in 'before' hooks in spec_helper.rb).

**Comment by movitto** [comment]
@roliveri let me know if there is anything else

File: spec/dir_class_spec.rb (hunk 31 above)
Comment (roliveri) [reply spec/dir_class_spec.rb:14]: @movitto Maybe the confusion is in the test doesn't seem to be testing what it states:
`"should enumerate the same file names as the standard Dir.glob - simple glob"`
It's not calling the standard `Dir.glob`, so how can it be testing that they return the same file names?

In the VirtFS tests, the mounted filesystem is NativeFS, so the same files and directories are visible through VirtFS as well as through the standard Ruby interface, so we can compare the return values. For Ext3, the files and directories ore only visible through VirtFS, so you can't compare the return values with the standard call.

**Review by roliveri — CHANGES_REQUESTED** [review]
@movitto I'm sorry, I'm finding it very hard to review the incremental changes of this PR. Is there any way you can restore the individual commits? I see absolutely no benefit in squashing commits, it just makes more work on both ends.

**Comment by movitto** [comment]
@roliveri I just started a new PR w/ everything split out in a manner that makes sense to the current state of the changes. I'm hoping it'll be easy to review & I got all your feedback.

The spec issue is still outstanding though, I believe this is just a misnomer as it isn't verifying the virtfs results against a runtime 'Dir.glob' but rather the glob output that I generated locally. We could probably rename these test cases to something like 'it should enumerate the same file names as the native ruby mechanisms' or similar.
